File: src/main/java/ua/softgroup/matrix/server/service/impl/TrackingDataServiceImpl.java
package ua.softgroup.matrix.server.service.impl;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import ua.softgroup.matrix.api.model.datamodels.ActiveWindowModel;
import ua.softgroup.matrix.server.persistent.entity.Project;
import ua.softgroup.matrix.server.persistent.entity.Screenshot;
import ua.softgroup.matrix.server.persistent.entity.TrackingData;
import ua.softgroup.matrix.server.persistent.entity.User;
import ua.softgroup.matrix.server.persistent.entity.WindowTime;
import ua.softgroup.matrix.server.persistent.entity.WorkDay;
import ua.softgroup.matrix.server.persistent.entity.WorkTimePeriod;
import ua.softgroup.matrix.server.persistent.repository.TrackingDataRepository;
import ua.softgroup.matrix.server.service.ProjectService;
import ua.softgroup.matrix.server.service.TrackingDataService;
import ua.softgroup.matrix.server.service.UserService;
import ua.softgroup.matrix.server.service.WorkDayService;
import ua.softgroup.matrix.server.service.WorkTimePeriodService;

import javax.imageio.ImageIO;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.InputStream;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;

@Service
public class TrackingDataServiceImpl extends AbstractEntityTransactionalService<TrackingData> implements TrackingDataService {

    private static final Logger logger = LoggerFactory.getLogger(TrackingDataServiceImpl.class);

    private static final String CWD = System.getProperty("user.dir");
    private static final String SCREENSHOT_EXT = "png";

    private final ProjectService projectService;
    private final UserService userService;
    private final WorkDayService workDayService;
    private final WorkTimePeriodService workTimePeriodService;

    @Autowired
    public TrackingDataServiceImpl(TrackingDataRepository repository, ProjectService projectService,
                                   UserService userService, WorkDayService workDayService,
                                   WorkTimePeriodService workTimePeriodService) {
        super(repository);
        this.projectService = projectService;
        this.userService = userService;
        this.workDayService = workDayService;
        this.workTimePeriodService = workTimePeriodService;
    }

    @Override
    @Transactional
    public void saveTrackingData(String userToken, Long projectId, String keyboardText, Double mouseFootage,
                                 List<ActiveWindowModel> activeWindowList, byte[] screenshot, String screenshotTitle) {

        logger.info("Saving tracking data: symbols {}, mouse {}, windows {}", keyboardText.length(), mouseFootage, activeWindowList.size());

        User user = userService.getByTrackerToken(userToken).orElseThrow(NoSuchElementException::new);
        Project project = projectService.getById(projectId).orElseThrow(NoSuchElementException::new);
        WorkDay workDay = workDayService.getByAuthorAndProjectAndDate(user, project, LocalDate.now()).orElseThrow(NoSuchElementException::new);
        WorkTimePeriod workTimePeriod = workTimePeriodService.getLatestPeriodOf(workDay).orElseThrow(NoSuchElementException::new);
        TrackingData trackingData = Optional.ofNullable(workTimePeriod.getTrackingData())
                                            .orElseGet(() -> new TrackingData(workTimePeriod));

//        TrackingData trackingData = getTrackingDataOf(userToken, projectId, LocalDate.now());
        trackingData.setKeyboardText(trackingData.getKeyboardText() + keyboardText);
        workDay.setSymbolsCount(workDay.getSymbolsCount() + keyboardText.length());
        trackingData.setMouseFootage(trackingData.getMouseFootage() + mouseFootage);
        if (screenshot != null) {

            try (InputStream is = new ByteArrayInputStream(screenshot)) {
                File screenshotFile = new File(CWD + "/screenshots/" + System.nanoTime() + "." + SCREENSHOT_EXT);
                screenshotFile.getParentFile().mkdirs();
                ImageIO.write(ImageIO.read(is), SCREENSHOT_EXT, screenshotFile);

                trackingData.getScreenshots().add(new Screenshot(LocalDateTime.now(), screenshotTitle, screenshotFile.getPath(), trackingData));
            } catch (Exception e) {
                logger.error("Failed to save screenshot", e);
            }
        }
        activeWindowList.stream()
                .map(entry -> new WindowTime(entry.getWindowTitle(), entry.getStartTime(), entry.getWorkingPeriodSeconds(), trackingData))
                .forEach(windowTime -> trackingData.getActiveWindows().add(windowTime));
        workDay.setWindowsSwitchedCount(workDay.getWindowsSwitchedCount() + activeWindowList.size());

        save(trackingData);
    }

//    private TrackingData getTrackingDataOf(String userToken, Long projectId, LocalDate date) {
//        User user = userService.getByTrackerToken(userToken).orElseThrow(NoSuchElementException::new);
//        Project project = projectService.getById(projectId).orElseThrow(NoSuchElementException::new);
//        WorkDay workDay = workDayService.getByAuthorAndProjectAndDate(user, project, date).orElseThrow(NoSuchElementException::new);
//        TrackingData trackingData = getRepository().findByWorkDay(workDay);
//        return trackingData != null ? trackingData : new TrackingData(workDay);
//    }

    @Override
    protected TrackingDataRepository getRepository() {
        return (TrackingDataRepository) repository;
    }
}

File: src/main/scala/ua/softgroup/matrix/server/service/impl/TrackingDataServiceImpl.scala
package ua.softgroup.matrix.server.service.impl

import java.io._
import java.time.{LocalDate, LocalDateTime, YearMonth}
import java.time.format.DateTimeFormatter
import java.util.Optional
import javax.imageio.ImageIO

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ua.softgroup.matrix.api.model.datamodels.ActiveWindowModel
import ua.softgroup.matrix.server.persistent.entity.{Screenshot, TrackingData, WindowTime}
import ua.softgroup.matrix.server.persistent.repository.TrackingDataRepository
import ua.softgroup.matrix.server.service._

import scala.util.{Failure, Try}
import scala.util.control.NonFatal

/**
  * @author Oleksandr Tyshkovets <sg.olexander@gmail.com>
  */
@Service
class TrackingDataServiceImpl @Autowired() (
                              repository: TrackingDataRepository,
                              projectService: ProjectService,
                              userService: UserService,
                              workDayService: WorkDayService,
                              workTimePeriodService: WorkTimePeriodService) extends TrackingDataService {

  private val logger = LoggerFactory.getLogger(this.getClass)

  private val SCREENSHOTS_FOLDER_PATH = System.getProperty("user.home") + "/screenshots"
  private val SCREENSHOT_EXT = "png"

  private val monthYearFormatter = DateTimeFormatter.ofPattern("yyyy-MM")

  @Transactional
  override def saveTrackingData(userToken: String, projectId: Long, keyboardText: String, mouseFootage: Double,
                                activeWindowList: List[ActiveWindowModel], screenshot: Array[Byte],
                                screenshotTitle: String): Unit = {

    logger.info(s"Saving tracking data: symbols ${keyboardText.length}, mouse $mouseFootage, windows ${activeWindowList.size}")

    val user = userService.getByTrackerToken(userToken).orElseThrow(() => new NoSuchElementException)
    val project = projectService.getById(projectId).orElseThrow(() => new NoSuchElementException)
    val workDay = workDayService.getByAuthorAndProjectAndDate(user, project, LocalDate.now).orElseThrow(() => new NoSuchElementException)
    val workTimePeriod = workTimePeriodService.getLatestPeriodOf(workDay).orElseThrow(() => new NoSuchElementException)
    val trackingData = Optional.ofNullable(workTimePeriod.getTrackingData).orElseGet(() => new TrackingData(workTimePeriod))

    trackingData.setKeyboardText(trackingData.getKeyboardText + keyboardText)
    trackingData.setMouseFootage(trackingData.getMouseFootage + mouseFootage)
    if (screenshot != null) {

      autoCloseResource(new ByteArrayInputStream(screenshot)) {
        is => {
          val filePath = s"$SCREENSHOTS_FOLDER_PATH/${monthYearFormatter.format(YearMonth.now)}/${System.nanoTime}.$SCREENSHOT_EXT"
          val screenshotFile = new File(filePath)
          screenshotFile.getParentFile.mkdirs
          ImageIO.write(ImageIO.read(is), SCREENSHOT_EXT, screenshotFile)

          trackingData.getScreenshots.add(new Screenshot(LocalDateTime.now, screenshotTitle, screenshotFile.getPath, trackingData))
        }
      } recover {
        case e: Exception => logger.error("Failed to save screenshot", e)
      }
    }

    activeWindowList
      .foreach(entry => {
        val windowTime = new WindowTime(entry.getWindowTitle, entry.getStartTime, entry.getWorkingPeriodSeconds, trackingData)
        trackingData.getActiveWindows.add(windowTime)
      })
    workDay.setWindowsSwitchedCount(workDay.getWindowsSwitchedCount + activeWindowList.size)
    workDay.setSymbolsCount(workDay.getSymbolsCount + keyboardText.length)

    save(trackingData)
  }

  override def getById(id: Long): Optional[TrackingData] = Optional.ofNullable(repository.findOne(id))

  override def save(entity: TrackingData): TrackingData = repository.save(entity)

  override def isExist(id: Long): Boolean = repository.exists(id)

  //TODO extract this method to separate class
  private def autoCloseResource[C <: Closeable, R](resource: => C)(fun: C => R): Try[R] =
    Try(resource).flatMap(resourceInstance => {
      try {
        val returnValue = fun(resourceInstance)
        Try(resourceInstance.close()).map(_ => returnValue)
      } catch {
        case NonFatal(exceptionInFunction) =>
          try {
            resourceInstance.close()
            Failure(exceptionInFunction)
          } catch {
            case NonFatal(exceptionInClose) =>
              exceptionInFunction.addSuppressed(exceptionInClose)
              Failure(exceptionInFunction)
          }
      }
    })

}